File: sdk/genius-games-py/genius_games/simulator.py
"""
Game simulator for running experiments and analysis
"""

import asyncio
import time
from typing import List, Dict, Any, Optional, Union
import logging
import statistics
from datetime import datetime

from .swarm import create_ai_swarm, AISwarm
from .types import GameConfig, GameType, SimulationResults


class GameSimulator:
    """Simulator for running game experiments"""
    
    def __init__(self, server_url: str, logger: Optional[logging.Logger] = None):
        """Initialize simulator"""
        self.server_url = server_url
        self.logger = logger or logging.getLogger("GameSimulator")
    
    async def run_simulation(
        self,
        games: List[Union[GameConfig, Dict[str, Any]]],
        swarm_size: int,
        swarm_type: str = "mixed",
        rounds: int = 10,
        verbose: bool = True
    ) -> SimulationResults:
        """Run a simulation with multiple games and rounds"""
        
        start_time = time.time()
        self.logger.info(f"Starting simulation: {len(games)} game types, {rounds} rounds each, {swarm_size} players")
        
        # Create results tracking
        all_game_results = []
        total_games = len(games) * rounds
        completed_games = 0
        failed_games = 0
        player_wins: Dict[str, int] = {}
        player_scores: Dict[str, List[int]] = {}
        game_durations: List[float] = []
        emergence_events = 0
        
        # Create AI swarm
        swarm = await create_ai_swarm(
            self.server_url,
            size=swarm_size,
            name_prefix=f"Sim{swarm_type.capitalize()}",
            decision_maker=swarm_type,
            debug=False
        )
        
        try:
            # Run simulations
            for game_config in games:
                # Extract game type for logging
                if isinstance(game_config, dict):
                    game_type = game_config.get("gameType", "unknown")
                else:
                    game_type = game_config.game_type
                
                self.logger.info(f"\nSimulating {game_type} - {rounds} rounds")
                
                for round_num in range(rounds):
                    if verbose:
                        self.logger.info(f"  Round {round_num + 1}/{rounds}")
                    
                    round_start = time.time()
                    
                    try:
                        # Run game
                        results = await swarm.create_and_play_game(game_config)
                        
                        if results:
                            completed_games += 1
                            result = results[0]  # All players get same result
                            
                            # Track game duration
                            game_durations.append(time.time() - round_start)
                            
                            # Update win tracking
                            if result.winners:
                                for winner in result.winners:
                                    player_wins[winner] = player_wins.get(winner, 0) + 1
                            elif result.winner:
                                player_wins[result.winner] = player_wins.get(result.winner, 0) + 1
                            
                            # Update score tracking
                            for player_id, score in result.final_scores.items():
                                if player_id not in player_scores:
                                    player_scores[player_id] = []
                                player_scores[player_id].append(score)
                            
                            # Check for emergence events
                            if self._check_emergence(result, swarm_size):
                                emergence_events += 1
                            
                            all_game_results.append(result)
                        else:
                            failed_games += 1
                    
                    except Exception as e:
                        self.logger.error(f"Game failed: {e}")
                        failed_games += 1
                    
                    # Brief pause between games
                    await asyncio.sleep(1)
            
            # Calculate final statistics
            win_rates = {}
            avg_scores = {}
            
            for player_id in player_scores:
                # Win rate
                wins = player_wins.get(player_id, 0)
                win_rates[player_id] = wins / completed_games if completed_games > 0 else 0
                
                # Average score
                scores = player_scores[player_id]
                avg_scores[player_id] = statistics.mean(scores) if scores else 0
            
            # Average game duration
            avg_duration = statistics.mean(game_durations) if game_durations else 0
            
            # Create results
            results = SimulationResults(
                total_games=total_games,
                completed_games=completed_games,
                failed_games=failed_games,
                win_rates=win_rates,
                avg_scores=avg_scores,
                avg_game_duration=avg_duration,
                emergence_events=emergence_events,
                metadata={
                    "swarm_size": swarm_size,
                    "swarm_type": swarm_type,
                    "duration_seconds": time.time() - start_time,
                    "games_per_second": completed_games / (time.time() - start_time) if completed_games > 0 else 0
                }
            )
            
            # Log summary
            self.logger.info("\n=== SIMULATION COMPLETE ===")
            self.logger.info(f"Total games: {total_games}")
            self.logger.info(f"Completed: {completed_games} ({completed_games/total_games*100:.1f}%)")
            self.logger.info(f"Failed: {failed_games}")
            self.logger.info(f"Average game duration: {avg_duration:.2f}s")
            self.logger.info(f"Emergence events: {emergence_events}")
            self.logger.info(f"Total time: {time.time() - start_time:.1f}s")
            
            if verbose:
                # Top performers
                self.logger.info("\nTop 5 by win rate:")
                top_winners = sorted(win_rates.items(), key=lambda x: x[1], reverse=True)[:5]
                for player, rate in top_winners:
                    self.logger.info(f"  {player}: {rate*100:.1f}%")
                
                self.logger.info("\nTop 5 by average score:")
                top_scorers = sorted(avg_scores.items(), key=lambda x: x[1], reverse=True)[:5]
                for player, score in top_scorers:
                    self.logger.info(f"  {player}: {score:.1f}")
            
            return results
            
        finally:
            # Disconnect swarm
            await swarm.disconnect_all()
    
    def _check_emergence(self, game_result: Any, swarm_size: int) -> bool:
        """Check if emergence behavior occurred"""
        # Simple heuristic: unusual concentration of winners
        if hasattr(game_result, 'winners') and game_result.winners:
            winner_ratio = len(game_result.winners) / swarm_size
            # Emergence if very few or very many winners (unusual consensus)
            return winner_ratio < 0.1 or winner_ratio > 0.9
        return False
    
    async def quick_test(self, game_type: GameType, players: int = 4) -> Dict[str, Any]:
        """Run a quick test of a single game"""
        self.logger.info(f"Quick test: {game_type} with {players} players")
        
        # Create small swarm
        swarm = await create_ai_swarm(
            self.server_url,
            size=players,
            name_prefix="QuickTest",
            decision_maker="mixed"
        )
        
        try:
            # Run one game
            config = {
                "gameType": game_type.value,
                "rounds": 10,
                "timeLimitMs": 5000
            }
            
            start = time.time()
            results = await swarm.create_and_play_game(config)
            duration = time.time() - start
            
            if results:
                result = results[0]
                return {
                    "success": True,
                    "duration": duration,
                    "winner": result.winner,
                    "winners": result.winners,
                    "final_scores": result.final_scores,
                    "rounds_played": result.rounds_played
                }
            else:
                return {
                    "success": False,
                    "error": "No results returned"
                }
                
        except Exception as e:
            return {
                "success": False,
                "error": str(e)
            }
        finally:
            await swarm.disconnect_all()
    
    async def benchmark_games(self, players: int = 8) -> Dict[str, Any]:
        """Benchmark all game types"""
        self.logger.info(f"Benchmarking all games with {players} players")
        
        benchmark_results = {}
        
        # Test each game type
        for game_type in GameType:
            self.logger.info(f"\nBenchmarking {game_type.value}...")
            
            try:
                result = await self.quick_test(game_type, players)
                benchmark_results[game_type.value] = result
                
                if result["success"]:
                    self.logger.info(f"✓ {game_type.value}: {result['duration']:.2f}s")
                else:
                    self.logger.info(f"✗ {game_type.value}: {result['error']}")
                
            except Exception as e:
                self.logger.error(f"✗ {game_type.value}: {str(e)}")
                benchmark_results[game_type.value] = {
                    "success": False,
                    "error": str(e)
                }
            
            # Brief pause
            await asyncio.sleep(1)
        
        # Summary
        successful = sum(1 for r in benchmark_results.values() if r.get("success", False))
        self.logger.info(f"\n=== BENCHMARK COMPLETE ===")
        self.logger.info(f"Successful: {successful}/{len(GameType)}")
        
        return benchmark_results

File: sdk/genius-games-js/src/simulator.ts
import { GeniusGamesClient } from './client';
import { AISwarm } from './swarm';
import { 
  ConnectionOptions,
  GameConfig,
  GameType,
  GameResult,
  PlayerInfo,
} from './types';

/**
 * Options for running a simulation
 */
export interface SimulationOptions {
  games: GameConfig[];
  swarmSize: number;
  swarmType?: 'random' | 'strategic' | 'mixed';
  rounds?: number;
  delayBetweenGames?: number;
  verbose?: boolean;
}

/**
 * Simulation results
 */
export interface SimulationResults {
  totalGames: number;
  completedGames: number;
  failedGames: number;
  results: GameResult[];
  winRates: Record<string, number>;
  avgGameDuration: number;
  avgScores: Record<string, number>;
  emergenceEvents: number;
}

/**
 * Game simulator for running experiments
 */
export class GameSimulator {
  private connectionOptions: ConnectionOptions;
  private humanClient?: GeniusGamesClient;

  constructor(connectionOptions: ConnectionOptions) {
    this.connectionOptions = connectionOptions;
  }

  /**
   * Run a simulation with AI players
   */
  async runSimulation(options: SimulationOptions): Promise<SimulationResults> {
    const results: SimulationResults = {
      totalGames: options.games.length * (options.rounds || 1),
      completedGames: 0,
      failedGames: 0,
      results: [],
      winRates: {},
      avgGameDuration: 0,
      avgScores: {},
      emergenceEvents: 0,
    };
    
    // Create AI swarm
    const swarm = new AISwarm(this.connectionOptions, {
      size: options.swarmSize,
      namePrefix: 'SimBot',
      decisionMaker: options.swarmType === 'mixed' ? 'random' : options.swarmType || 'random',
    });
    
    try {
      await swarm.connect();
      
      if (options.verbose) {
        console.log(`Starting simulation with ${options.swarmSize} AI players`);
      }
      
      // Run games
      for (let round = 0; round < (options.rounds || 1); round++) {
        for (const gameConfig of options.games) {
          if (options.verbose) {
            console.log(`\nRound ${round + 1}: Starting ${gameConfig.gameType}`);
          }
          
          try {
            const gameResult = await this.runSingleGame(swarm, gameConfig, options);
            results.results.push(gameResult);
            results.completedGames++;
            
            // Update statistics
            this.updateStats(results, gameResult);
            
            if (options.verbose) {
              console.log(`Game completed. Winner: ${gameResult.winner}`);
            }
          } catch (error) {
            results.failedGames++;
            if (options.verbose) {
              console.error('Game failed:', error);
            }
          }
          
          // Delay between games
          if (options.delayBetweenGames) {
            await new Promise(resolve => setTimeout(resolve, options.delayBetweenGames));
          }
        }
      }
      
      // Calculate final statistics
      this.calculateFinalStats(results);
      
    } finally {
      swarm.disconnect();
    }
    
    return results;
  }

  /**
   * Run a single game and observe as human
   */
  async observeGame(gameConfig: GameConfig, swarmSize: number): Promise<GameResult> {
    // Create human observer client
    const playerInfo: PlayerInfo = {
      id: 'human_observer',
      name: 'Observer',
      type: 'human',
    };
    
    this.humanClient = new GeniusGamesClient(this.connectionOptions, playerInfo);
    
    // Create AI swarm
    const swarm = new AISwarm(this.connectionOptions, {
      size: swarmSize,
      namePrefix: 'AI',
      decisionMaker: 'strategic',
    });
    
    try {
      // Connect everyone
      await Promise.all([
        this.humanClient.connect(),
        swarm.connect(),
      ]);
      
      // Create game
      const gameId = await this.humanClient.createGame(gameConfig);
      console.log(`Created game ${gameId}`);
      
      // Setup observation handlers
      return new Promise<GameResult>((resolve, reject) => {
        this.humanClient!.setHandlers({
          onGameStateUpdate: (state) => {
            console.log(`\nRound ${state.round}`);
            console.log('Scores:', state.scores);
          },
          
          onRoundResult: (result) => {
            console.log('Round result:', result.outcome.specialEvents);
            if (result.outcome.emergenceDetected) {
              console.log('🌟 Emergence detected!');
            }
          },
          
          onGameEnded: (result) => {
            console.log('\nGame ended!');
            console.log('Winner:', result.winner);
            console.log('Final scores:', result.finalScores);
            resolve(result);
          },
          
          onError: (error) => {
            reject(error);
          },
        });
        
        // AI swarm joins the game
        swarm.joinGame(gameId);
      });
      
    } finally {
      if (this.humanClient) {
        this.humanClient.disconnect();
      }
      swarm.disconnect();
    }
  }

  private async runSingleGame(
    swarm: AISwarm,
    gameConfig: GameConfig,
    options: SimulationOptions
  ): Promise<GameResult> {
    return new Promise<GameResult>((resolve, reject) => {
      const timeout = setTimeout(() => {
        reject(new Error('Game timeout'));
      }, 300000); // 5 minute timeout
      
      // Create temporary client to monitor game
      const monitorInfo: PlayerInfo = {
        id: 'monitor',
        name: 'Monitor',
        type: 'ai',
      };
      
      const monitor = new GeniusGamesClient(this.connectionOptions, monitorInfo);
      
      monitor.setHandlers({
        onGameEnded: (result) => {
          clearTimeout(timeout);
          monitor.disconnect();
          resolve(result);
        },
        
        onError: (error) => {
          clearTimeout(timeout);
          monitor.disconnect();
          reject(error);
        },
      });
      
      monitor.connect().then(() => {
        swarm.createAndPlayGame(gameConfig).catch(reject);
      }).catch(reject);
    });
  }

  private updateStats(results: SimulationResults, gameResult: GameResult): void {
    // Update win rates
    if (!results.winRates[gameResult.winner]) {
      results.winRates[gameResult.winner] = 0;
    }
    results.winRates[gameResult.winner]++;
    
    // Update average scores
    for (const [player, score] of Object.entries(gameResult.finalScores)) {
      if (!results.avgScores[player]) {
        results.avgScores[player] = 0;
      }
      results.avgScores[player] += score;
    }
    
    // Count emergence events
    results.emergenceEvents += gameResult.emergenceEvents.length;
    
    // Update average game duration
    results.avgGameDuration += gameResult.totalRounds;
  }

  private calculateFinalStats(results: SimulationResults): void {
    // Convert win counts to rates
    for (const player in results.winRates) {
      results.winRates[player] = results.winRates[player] / results.completedGames;
    }
    
    // Calculate average scores
    for (const player in results.avgScores) {
      results.avgScores[player] = results.avgScores[player] / results.completedGames;
    }
    
    // Calculate average game duration
    if (results.completedGames > 0) {
      results.avgGameDuration = results.avgGameDuration / results.completedGames;
    }
  }
}

/**
 * Run a quick simulation
 */
export async function quickSimulation(
  serverUrl: string,
  gameType: GameType,
  numPlayers: number = 4,
  numGames: number = 10
): Promise<SimulationResults> {
  const simulator = new GameSimulator({ url: serverUrl });
  
  return simulator.runSimulation({
    games: [{
      gameType,
      rounds: 20,
      timeLimitMs: 5000,
    }],
    swarmSize: numPlayers,
    swarmType: 'mixed',
    rounds: numGames,
    verbose: true,
  });
}